### PR TITLE
Fix implementation of index_put, validate it with dort | fix(torchlib)

### PR DESCRIPTION
**Pull request by xadupre**
The onnx implementation of index_put is different in torch script exporter (https://github.com/pytorch/pytorch/blob/main/torch/onnx/symbolic_opset11.py#L212). The PR replaces the current implementation failing on one corner case by the one from torch script.

### CODE REVIEW

File: onnxscript/function_libs/torch_lib/ops/core.py [hunk 1]
```diff
@@ -4046,25 +4046,17 @@ def aten_index_put(
...
-    index = op.SequenceAt(indices, 0)  # assume indices only have 1 element
-    # change array([1,3]) to array([[1,1,1,1,1],[3,3,3,3,3]])
-    self_dim_1 = op.Gather(op.Shape(self), 1)
-    index_dim_0 = op.Gather(op.Shape(index), 0)
-    neg_1 = op.Constant(value_ints=[-1])
-    shape = op.Concat(op.Reshape(self_dim_1, neg_1), op.Reshape(index_dim_0, neg_1), axis=0)
-    new_ind = op.Expand(index, shape)
-    new_ind_t = op.Transpose(new_ind)
+    index = op.SequenceAt(indices, 0)
+    new_index = op.Unsqueeze(index, -1)
```
Comment (gramalingam) [inline onnxscript/function_libs/torch_lib/ops/core.py:4050]: nit: I wonder if `Unsqueeze(index, [-1])` is more precise. My understanding is that the second argument should be a 1-dimensional tensor, but `-1` will produce a scalar -1. Even if it works with ORT, it may be better to stick closer to expectation.

File: onnxscript/tests/function_libs/torch_lib/ops_test_data.py [hunk 2]
```diff
@@ -836,14 +836,18 @@ def _where_input_wrangler(
...
         core_ops.aten_index_put,
     ).skip(
-        matcher=lambda sample: not (sample.args[0][0].dtype == torch.int64),
-        reason="this Aten overload only support tensor(int) as args",
+        matcher=lambda sample: not (
+            (sample.args[0][0].dtype == torch.int64)
+            # onnxruntime: MLFloat16 data type is not supported with ScatterND when reduction is 'add'
+            and (sample.args[1].dtype != torch.float16 or not sample.kwargs.get("accumulate", False))
+        ),
+        reason="this Aten overload only support tensor(int) as indices or float16 when reduction is 'add'",
```
Comment (titaiwangms) [inline onnxscript/tests/function_libs/torch_lib/ops_test_data.py:850]: I tested this matcher with the original code and it shows no different in results (still 4 passes and 20 skipped). It seems that you found a corner case that is not in the current `op_DB_test`? If that's the case, we usually use https://github.com/microsoft/onnxscript/blob/main/onnxscript/tests/function_libs/torch_lib/extra_opinfo.py to create corner case to test the implementation.

**Comment by titaiwangms** [comment]
Please also update the PR description so that torchlib team: @justinchuby @xiaowuhu @fatcat-z can have more context later. Thanks!

File: onnxscript/tests/function_libs/torch_lib/dynamo_export_test.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+        values = torch.arange(24, dtype=torch.float32).reshape((-1, 4, 3))
+
+        # redondant test to make sure this expression is valid for torch
+        assert x.index_put((indices,), values) is not None
+
+        self.assertONNX(
+            lambda x, indices, values: x.index_put((indices,), values, accumulate=True),
+            (x, indices, values),
+            onnx_export=inspect.currentframe().f_code.co_name,
+        )
```
[the remark this answers is not in the document]
Comment (titaiwangms) [reply onnxscript/tests/function_libs/torch_lib/dynamo_export_test.py:173]: The README.md might be not informative. You can reference on [this PR](https://github.com/microsoft/onnxscript/pull/1246/files) of how to create a test in `extra_opinfo.py` and enables it in `ops_test_data.py`.

[the remark this answers is not in the document]
Comment (justinchuby) [reply onnxscript/tests/function_libs/torch_lib/dynamo_export_test.py:173]: @titaiwangms looks like we can integrate the "test backward" part to the integration test you have in mind to test backward computation as well.

@xadupre I would recommend the same route for testing as Ti-Tai suggested. Thanks!

File: onnxscript/tests/function_libs/torch_lib/ops_test_data.py [hunk 4]
```diff
@@ -836,14 +836,18 @@ def _where_input_wrangler(
         core_ops.aten_index_put_bool,
     ).skip(
         matcher=lambda sample: not (sample.args[0][0].dtype == torch.bool),
-        reason="this Aten overload only support tensor(bool) as args",
+        reason="this Aten overload only support tensor(bool) as indices",
     ),
     TorchLibOpInfo(
         "index_put",
         core_ops.aten_index_put,
     ).skip(
```
Comment (justinchuby) [inline onnxscript/tests/function_libs/torch_lib/ops_test_data.py:844]: Is xfail possible?

File: onnxscript/function_libs/torch_lib/ops/core.py [hunk 5]
```diff
@@ -4046,25 +4046,17 @@ def aten_index_put(
...
 
-    index = op.SequenceAt(indices, 0)  # assume indices only have 1 element
-    # change array([1,3]) to array([[1,1,1,1,1],[3,3,3,3,3]])
-    self_dim_1 = op.Gather(op.Shape(self), 1)
-    index_dim_0 = op.Gather(op.Shape(index), 0)
-    neg_1 = op.Constant(value_ints=[-1])
-    shape = op.Concat(op.Reshape(self_dim_1, neg_1), op.Reshape(index_dim_0, neg_1), axis=0)
-    new_ind = op.Expand(index, shape)
-    new_ind_t = op.Transpose(new_ind)
+    index = op.SequenceAt(indices, 0)
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_lib/ops/core.py:4049]: I would add a reference and/or a more detailed description of the implementation for this op / key lines since it is rather complex to get right.

Comment (xadupre) [reply onnxscript/function_libs/torch_lib/ops/core.py:4049]: It is a copy/paste from torch.onnx. I can add a link to the sources.

File: onnxscript/function_libs/torch_lib/ops/core.py (hunk 1 above)
Comment (xadupre) [reply onnxscript/function_libs/torch_lib/ops/core.py:4050]: done

File: onnxscript/function_libs/torch_lib/ops/core.py (hunk 5 above)
Comment (xadupre) [reply onnxscript/function_libs/torch_lib/ops/core.py:4049]: done

File: onnxscript/tests/function_libs/torch_lib/ops_test_data.py (hunk 2 above)
Comment (xadupre) [reply onnxscript/tests/function_libs/torch_lib/ops_test_data.py]: I did. I hope I did it right.

File: onnxscript/tests/function_libs/torch_lib/dynamo_export_test.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (xadupre) [reply onnxscript/tests/function_libs/torch_lib/dynamo_export_test.py:173]: I did not understand how the backward verification is triggered. Do I need to add more than the corner case?

File: onnxscript/tests/function_libs/torch_lib/ops_test_data.py (hunk 4 above)
Comment (xadupre) [reply onnxscript/tests/function_libs/torch_lib/ops_test_data.py:844]: They may succeed with a newer version of onnxruntime.

File: onnxscript/tests/function_libs/torch_lib/ops_test_data.py [hunk 6]
```diff
@@ -836,14 +836,21 @@ def _where_input_wrangler(
...
+            (sample.args[0][0].dtype == torch.int64)
+            # onnxruntime: MLFloat16 data type is not supported with ScatterND when reduction is 'add'
+            and (
+                sample.args[1].dtype != torch.float16
+                or not sample.kwargs.get("accumulate", False)
+            )
+        ),
+        reason="this Aten overload only support tensor(int) as indices and float32 when accumulate is True",
     ),
     TorchLibOpInfo("index_select", core_ops.aten_index_select),
```
Comment (justinchuby) [inline onnxscript/tests/function_libs/torch_lib/ops_test_data.py:855]:
```suggestion
    TorchLibOpInfo("ops.aten.index_put", core_ops.aten_index_put),
    TorchLibOpInfo("index_select", core_ops.aten_index_select),
```

We then need to enable the test case like so

File: onnxscript/tests/function_libs/torch_lib/extra_opinfo.py [hunk 7]
```diff
@@ -741,6 +741,28 @@ def sample_inputs_index(op_info, device, dtype, requires_grad, **kwargs):
...
+def sample_inputs_index_put(op_info, device, dtype, requires_grad, **kwargs):
+    del op_info
+    del kwargs
+
+    data = torch_testing.make_tensor(
+        (10, 3),
+        device=device,
+        dtype=dtype,
+        requires_grad=requires_grad,
+        **kwargs,
```
Comment (justinchuby) [inline onnxscript/tests/function_libs/torch_lib/extra_opinfo.py:753]:
```suggestion
```

File: onnxscript/tests/function_libs/torch_lib/dynamo_export_test.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (justinchuby) [reply onnxscript/tests/function_libs/torch_lib/dynamo_export_test.py:173]: Just the corner case should suffice. When we are certain the cases are covered iextra_opinfo.py`, feel free to remove this test file.

File: onnxscript/tests/function_libs/torch_lib/ops_test_data.py (hunk 4 above)
Comment (justinchuby) [reply onnxscript/tests/function_libs/torch_lib/ops_test_data.py:844]: You may use the `enabled_if` option: https://github.com/microsoft/onnxscript/blob/457e52e10ef4f16e471cebf8e21c1e8fad012219/onnxscript/tests/function_libs/torch_lib/ops_test_data.py#L553

**Comment by titaiwangms** [comment]
Little changes got merge in main. Might be unrelated/related.

File: onnxscript/tests/function_libs/torch_lib/dynamo_export_test.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (justinchuby) [reply onnxscript/tests/function_libs/torch_lib/dynamo_export_test.py:173]: Backed up in https://gist.github.com/justinchuby/b9d0809667b2bbaacd574d5de00291f5

File: onnxscript/tests/function_libs/torch_lib/ops_test_data.py [hunk 8]
```diff
@@ -836,14 +836,21 @@ def _where_input_wrangler(
         core_ops.aten_index_put_bool,
     ).skip(
         matcher=lambda sample: not (sample.args[0][0].dtype == torch.bool),
-        reason="this Aten overload only support tensor(bool) as args",
+        reason="this Aten overload only support tensor(bool) as indices",
     ),
     TorchLibOpInfo(
         "index_put",
         core_ops.aten_index_put,
     ).skip(
```
Comment (justinchuby) [inline onnxscript/tests/function_libs/torch_lib/ops_test_data.py:844]:
```suggestion
    ).skip(
    enabled_if=version_utils.onnxruntime_older_than("1.16"), 
```

File: onnxscript/tests/function_libs/torch_lib/ops_test_data.py [hunk 9]
```diff
@@ -836,15 +836,24 @@ def _where_input_wrangler(
...
-        matcher=lambda sample: not (sample.args[0][0].dtype == torch.int64),
-        reason="this Aten overload only support tensor(int) as args",
+        enabled_if=version_utils.onnxruntime_older_than("1.16"),
```
Comment (justinchuby) [inline onnxscript/tests/function_libs/torch_lib/ops_test_data.py:845]:
```suggestion
        enabled_if=version_utils.onnxruntime_older_than("1.17"), 
```

File: onnxscript/function_libs/torch_lib/ops/core.py [hunk 10]
```diff
@@ -4044,27 +4044,22 @@ def aten_index_put(
...
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_lib/ops/core.py:4052]:
```suggestion
    TODO(justinchuby): Handle when indicies has more than one element
    index = op.SequenceAt(indices, 0)
```

File: onnxscript/function_libs/torch_lib/ops/core.py [hunk 11]
```diff
@@ -4044,27 +4044,23 @@ def aten_index_put(
...
 
-    index = op.SequenceAt(indices, 0)  # assume indices only have 1 element
-    # change array([1,3]) to array([[1,1,1,1,1],[3,3,3,3,3]])
-    self_dim_1 = op.Gather(op.Shape(self), 1)
-    index_dim_0 = op.Gather(op.Shape(index), 0)
-    neg_1 = op.Constant(value_ints=[-1])
-    shape = op.Concat(op.Reshape(self_dim_1, neg_1), op.Reshape(index_dim_0, neg_1), axis=0)
-    new_ind = op.Expand(index, shape)
-    new_ind_t = op.Transpose(new_ind)
+    TODO(justinchuby): Handle when indicies has more than one element
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_lib/ops/core.py:4052]:
```suggestion
    # TODO(justinchuby): Handle when indicies has more than one element
```

**Comment by justinchuby** [comment]
Merging into a microsoft branch so that it is easier to work on.
